### PR TITLE
add support for migration v2 to the spec

### PR DESCRIPTION
**Pull request by w-henderson**
This adds the VFIO_USER_DEVICE_FEATURE, VFIO_USER_MIG_DATA_READ, and VFIO_USER_MIG_DATA_WRITE commands.

### CODE REVIEW

**Review by tmakatos — COMMENTED** [review]
Which kernel version is this based on?

File: docs/vfio-user.rst [hunk 1]
```diff
@@ -515,18 +518,7 @@ Capabilities:
...
-| Name            | Type   | Description                                      |
-+=================+========+==================================================+
-| pgsize          | number | Page size of dirty pages bitmap. The smallest    |
-|                 |        | between the client and the server is used.       |
-+-----------------+--------+--------------------------------------------------+
-| max_bitmap_size | number | Maximum bitmap size in ``VFIO_USER_DIRTY_PAGES`` |
-|                 |        | and ``VFIO_DMA_UNMAP`` messages.  Optional,      |
-|                 |        | with a default value of 256MB.                   |
-+-----------------+--------+--------------------------------------------------+
+The migration capability is currently an empty object.
```
Comment (tmakatos) [inline docs/vfio-user.rst:521]: Can you explain in the commit message why this isn't needed anymore?

File: docs/vfio-user.rst [hunk 2]
```diff
@@ -1468,6 +1460,393 @@ Reply
...
+
+Direct State Transitions
+""""""""""""""""""""""""
+
+The device migration FSM is a Mealy machine, so actions are taken upon the arcs
+between FSM states. The following transitions need to be supported by the
+server, a subset of those defined in ``<linux/vfio.h>``
+(``enum vfio_device_mig_state``).
+
+* ``RUNNING -> STOP``, ``STOP_COPY -> STOP``: Stop the operation of the device. The ``STOP_COPY`` arc terminates the data transfer session.
```
Comment (tmakatos) [inline docs/vfio-user.rst:1717]: nit: line should be < 80

**Review by tmakatos — COMMENTED** [review]
I haven't looked at the content yet, will do tomorrow.

File: docs/vfio-user.rst (hunk 1 above)
Comment (w-henderson) [reply docs/vfio-user.rst]: Will do.

File: docs/vfio-user.rst [hunk 3]
```diff
@@ -1468,6 +1455,396 @@ Reply
...
+  if the internal page size is lower than requested
+
+The data field of the response is identical, except with the bitmap added on
+the end at offset 24.
+
+The mapping of IOVA to bits is given by:
+
+``bitmap[(addr - iova)/page_size] & (1ULL << (addr % 64))``
+
+Direct State Transitions
```
Comment (jlevon) [inline docs/vfio-user.rst:1706]: shouldn't this and next be a sub-section of the DEVICE_STATE section?

File: docs/vfio-user.rst [hunk 4]
```diff
@@ -1468,6 +1455,396 @@ Reply
...
++-------+--------+--------------------------------+
+| data  | 8      | variable                       |
++-------+--------+--------------------------------+
+
+* *argsz* is the size of the above structure.
+
+* *flags* defines the action to be performed by the server and upon which
+  feature:
+
+  * The feature is selected using the 16-bit ``VFIO_DEVICE_FEATURE_MASK``
```
[the remark this answers is not in the document]
Comment (w-henderson) [reply docs/vfio-user.rst:1496]: How about "the feature bits are the least significant 16 bits of the flags field, and can be accessed using the ``VFIO_DEVICE_FEATURE_MASK`` bit mask."?

File: docs/vfio-user.rst [hunk 5]
```diff
@@ -1468,6 +1455,396 @@ Reply
...
+    portion of the flags field.
```
[the remark this answers is not in the document]
Comment (w-henderson) [reply docs/vfio-user.rst:1497]: > ``VFIO_DEVICE_FEATURE_GET`` and ``VFIO_DEVICE_FEATURE_SET`` are mutually exclusive, except for use with ``VFIO_DEVICE_FEATURE_PROBE``.

Or am I misunderstanding your question?

File: docs/vfio-user.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (jlevon) [reply docs/vfio-user.rst]: :+1:

**Review by jlevon — APPROVED** [review]

**Comment by tmakatos** [comment]
@w-henderson I think this can now be merged into [master-migv2](https://github.com/nutanix/libvfio-user/tree/master-migv2)?

**Comment by w-henderson** [comment]
@tmakatos Yes, I think it's all good - go ahead!

**Comment by tmakatos** [comment]
@w-henderson looks like there are conflicts?

**Comment by w-henderson** [comment]
@tmakatos I have no idea what was going on with the conflicts - Git seemed to be finding conflicts between identical paragraphs of text and the above merge commit is literally an empty commit, but apparently that fixed the conflicts? Anyway it's good to merge now (I hope).

**Review by tmakatos — APPROVED** [review]

**Comment by tmakatos** [comment]
Ugh, I botched the commit message. Anyway, let's fix it when merging to master.

**Pull request merged**